### PR TITLE
i#3230 split traces: add parallel trace analysis support

### PR DESCRIPTION
**Pull request by derekbruening**
Adds support for analyzing traces in parallel, concurrently operating on each
traced thread.  This is made possible by the new storage of traces in separate
per-thread files.

Adds a new analysis_tool_t interface where if the tool's
parallel_shard_supported() returns true, analyzer_t switches to a parallel
operation mode.  Today, a simple static scheduling among worker threads is used.
Each worker completely owns one or more shards, eliminating the need for
synchronization when processing a shard's trace entries.  The default shard is
today's trace file split, a traced thread.

A tool's parallel_shard_init() function is invoked to create shard-thread-local
data, which is passed to parallel_shard_memref().  Errors are also shard-local
with parallel_shard_error().  A parallel_shard_exit() is provided for cleanup,
though most tools will sort, aggregate, and clean up in print_results().

Implements the new interface in the basic_counts and opcode_mix tools.  More
tools will be converted in the future.

Adds a new routine module_mapper_t::find_mapped_trace_bounds() so opcode_mix can
perform local caching of mappings, to avoid a global lock bottleneck for
module_mapper_t usage.

Issue: #3230

### CODE REVIEW

File: clients/drcachesim/analyzer.cpp [hunk 1]
```diff
@@ -39,43 +40,83 @@
...
+    if (parallel && directory_iterator_t::is_directory(trace_path)) {
+        directory_iterator_t end;
+        directory_iterator_t iter(trace_path);
+        if (!iter) {
+            ERRMSG("Failed to list directory %s: %s", trace_path.c_str(),
+                   iter.error_string().c_str());
+            return false;
+        }
+        for (; iter != end; ++iter) {
+            std::string fname = *iter;
```
Comment (snehasish) [inline clients/drcachesim/analyzer.cpp]: const std::string

File: clients/drcachesim/analyzer.cpp [hunk 2]
```diff
@@ -39,43 +40,83 @@
...
+            if (fname == "." || fname == "..")
+                continue;
+            std::string path = trace_path + DIRSEP + fname;
```
Comment (snehasish) [inline clients/drcachesim/analyzer.cpp]: const std::string

**Comment by derekbruening** [comment]
> Design alternative to consider:
> Shard data and worker data is referenced via void*. What do you think about templating analysis_tool_t for shard data and worker data? With this change we can return a std::unique_ptr from parallel_shard_init which would allow analysis_tool_t to own the objects. Thus analysis tools no longer have to worry about freeing memory of shard data and worker data in their destructor and corresponding exit functions. Right now tool writers need to reason about things like "can parallel_shard_init be called after parallel_worker_exit". If all of the data was owned by the analyzer I think it would simplify analysis tools. However, this would change the way analyzer::process_tasks() is implemented right now.

I think you mean `s/allow analysis_tool_t to own/allow analyzer_t to own/`.  But I'm not sure I follow: do you mean templating analyzer_t, specializing it for just one type of tool?

File: clients/drcachesim/analysis_tool.h [hunk 3]
```diff
@@ -48,13 +48,39 @@
...
+ * class.
+ *
+ * Concurrent processing of traces is supported by logically splitting a trace into
+ * "shards" which are each processed sequentially.  The default shard is a traced
+ * application thread, but the tool interface can support other divisions.  For tools
+ * that support concurrent processing of shards and do not need to see a single
+ * time-sorted interleaved merged trace, the interface functions with the parallel_
+ * prefix should be implemented, and parallel_shard_supported() should return true.
+ * parallel_shard_init() will be invoked for each shard prior to invoking
+ * parallel_shard_memref() for each entry in that shard; the data structure returned
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/analysis_tool.h]: Done.

File: clients/drcachesim/analysis_tool.h [hunk 4]
```diff
@@ -103,6 +129,86 @@ class analysis_tool_t {
     virtual bool
     print_results() = 0;
 
+    /**
+     * Returns whether this tool supports analyzing trace shards concurrently, or
+     * whether it needs to see a single time-sorted interleaved stream of traced
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/analysis_tool.h]: The traced events are interleaved in some order, and that order is chronological by buffer timestamp.  Either there is no mention of ordering, or there is a long drawn-out discussion on 32K buffers being much smaller than scheduling time quanta and so being accurate for non-memory-sharing threads but not perfect for memory-sharing interleavings, or there is a high-level ordering description which this is meant to be.  IMHO it is adding more value than confusion.  But I am ok removing it.

File: clients/drcachesim/analysis_tool.h [hunk 5]
```diff
@@ -103,6 +129,86 @@ class analysis_tool_t {
...
+    /**
+     * Invoked once when all trace entries for a shard have been processed.  \p
+     * shard_data is the value returned by parallel_shard_init() for this shard.
+     * This allows a tool to clean up its thread data, or to report thread analysis
+     * results.  Most tools, however, prefer to aggregate data or at least sort data,
+     * and perform nothing here, doing all cleanup in print_results() by storing the
+     * thread data into a table.
+     */
+    virtual void
+    parallel_shard_exit(void *shard_data)
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/analysis_tool.h]: Done.

File: clients/drcachesim/analysis_tool.h [hunk 6]
```diff
@@ -103,6 +129,86 @@ class analysis_tool_t {
...
+    {
+    }
+    /**
+     * The heart of an analysis tool, this routine operates on a single trace entry
+     * and takes whatever actions the tool needs to perform its analysis. The \p
+     * shard_data parameter is the value returned by parallel_shard_init() for this
+     * shard.  Since each shard is operated upon in its entirety by the same worker
+     * thread, no synchronization is needed.  The return value indicates whether this
+     * function was successful. On failure, parallel_shared_error() returns a
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/analysis_tool.h]: Done.

File: clients/drcachesim/analyzer.cpp (hunk 1 above)
Comment (derekbruening) [reply clients/drcachesim/analyzer.cpp]: Done.

File: clients/drcachesim/analyzer.cpp (hunk 2 above)
Comment (derekbruening) [reply clients/drcachesim/analyzer.cpp]: Done.

File: clients/drcachesim/analyzer.cpp [hunk 7]
```diff
@@ -39,43 +40,83 @@
...
+                std::unique_ptr<reader_t>(new my_file_reader_t(path, verbosity)), path));
+        }
+        // Like raw2trace, we use a simple round-robin static work assigment.  This
+        // could be improved later with dynamic work queue for better load balancing.
+        if (worker_count <= 0)
+            worker_count = std::thread::hardware_concurrency();
+        worker_tasks.resize(worker_count);
+        int worker = 0;
+        for (size_t i = 0; i < thread_data.size(); ++i) {
+            VPRINT(this, 2, "Worker %d assigned trace thread %zd\n", worker, i);
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/analyzer.cpp]: Done.

File: clients/drcachesim/analyzer.cpp [hunk 8]
```diff
@@ -128,32 +169,96 @@ analyzer_t::get_error_string()
...
+            tdata->error = "Failed to read from trace" + tdata->trace_file;
+            return;
+        }
+        std::vector<void *> shard_data(num_tools);
+        for (int i = 0; i < num_tools; ++i)
+            shard_data[i] = tools[i]->parallel_shard_init(tdata->index, worker_data[i]);
+        VPRINT(this, 1, "shard_data[0] is %p\n", shard_data[0]);
+        for (; *tdata->iter != *trace_end; ++(*tdata->iter)) {
+            for (int i = 0; i < num_tools; ++i) {
+                memref_t memref = **tdata->iter;
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/analyzer.cpp]: Done.

File: clients/drcachesim/tools/basic_counts.h [hunk 9]
```diff
@@ -41,41 +41,79 @@
...
-    std::unordered_map<memref_tid_t, int_least64_t> thread_sched_markers;
-    std::unordered_map<memref_tid_t, int_least64_t> thread_xfer_markers;
-    std::unordered_map<memref_tid_t, int_least64_t> thread_func_id_markers;
-    std::unordered_map<memref_tid_t, int_least64_t> thread_func_retaddr_markers;
-    std::unordered_map<memref_tid_t, int_least64_t> thread_func_arg_markers;
-    std::unordered_map<memref_tid_t, int_least64_t> thread_func_retval_markers;
-    std::unordered_map<memref_tid_t, int_least64_t> thread_other_markers;
+    struct counters_t {
+        counters_t()
+            : tid(0)
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tools/basic_counts.h]: Done.

File: clients/drcachesim/tools/opcode_mix.cpp [hunk 10]
```diff
@@ -78,24 +79,83 @@ opcode_mix_t::initialize()
...
+    shard_data_t *shard = reinterpret_cast<shard_data_t *>(shard_data);
+    ++shard->instr_count;
+
     app_pc mapped_pc;
-    mapped_pc = module_mapper->find_mapped_trace_address((app_pc)memref.instr.addr);
-    if (!module_mapper->get_last_error().empty()) {
-        error_string = "Failed to find mapped address for " +
-            to_hex_string(memref.instr.addr) + ": " + module_mapper->get_last_error();
-        return false;
+    app_pc trace_pc = reinterpret_cast<app_pc>(memref.instr.addr);
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tools/opcode_mix.cpp]: Done.

File: clients/drcachesim/tools/opcode_mix.h [hunk 11]
```diff
@@ -43,28 +43,82 @@
...
+        worker_data_t *worker;
+        int_least64_t instr_count;
+        std::unordered_map<int, int_least64_t> opcode_counts;
+        std::string error;
+        app_pc last_trace_module_start;
+        size_t last_trace_module_size;
+        app_pc last_mapped_module_start;
+    };
+
+    struct scoped_mutex_t {
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tools/opcode_mix.h]: I think I will leave it as a down payment toward perhaps sharing it among multiple C++ apps in the future.  I will add such a comment.

**Comment by derekbruening** [comment]
> Can we add a test which exercises the parallel analysis for basic counts?

There are 5 existing tests that run basic_counts.  3 use single-threaded traced apps, but 2 have multiple threads and so exercise parallel analysis (since it's on by default): tool.drcacheoff.burst_threadfilter and tool.drcacheoff.burst_threadL0filter.  Or did you have some other type of test in mind?

**Comment by derekbruening** [comment]
PTAL

**Comment by snehasish** [comment]
> > Design alternative to consider:
> > Shard data and worker data is referenced via void*. What do you think about templating analysis_tool_t for shard data and worker data? With this change we can return a std::unique_ptr from parallel_shard_init which would allow analysis_tool_t to own the objects. Thus analysis tools no longer have to worry about freeing memory of shard data and worker data in their destructor and corresponding exit functions. Right now tool writers need to reason about things like "can parallel_shard_init be called after parallel_worker_exit". If all of the data was owned by the analyzer I think it would simplify analysis tools. However, this would change the way analyzer::process_tasks() is implemented right now.
> 
> I think you mean `s/allow analysis_tool_t to own/allow analyzer_t to own/`. But I'm not sure I follow: do you mean templating analyzer_t, specializing it for just one type of tool?

Yes, I think I meant analyzer_t. I think we can pass on this suggestion for now. To clarify, I was thinking about a better ownership model for shard and worker data. However templating analyzer_t for each tool makes it at odds with the current approach of multiple tools run by the same analyzer.

**Pull request merged**